### DISCUSSION TITLE
How to use ILogger.BeginScope in Blazor Server for the entire activity?

### DISCUSSION THREAD

**Discussion by alexandrejobin** [dotnet/aspnetcore#55949] [Q&A]
In a standard request/response app with AspNet.Core, it was easy to add informations in the request context so that every single log entry could have access to them automatically. For exemple, if I need that all my log entries have the FormId that he is presently editing, I just had to create a new **Middleware** that use **ILogger.BeginScope** to achieve that and register the middleware at the beginning of the pipeline.

How can we achieve the same thing in Blazor? The only way i've found is with NET 8 and the new CreateInboundActivityHandler that you can override in the CircuitHandler. Is it the way to go? Because depending of the object that you inject, the information inside is not always up-to-date. For exemple, NavigationManager.Uri doesn't have the new value yet and it return the uri where you come from and not the current one.

For exemple, 

```c#
public class LoggingCircuitHandler : CircuitHandler
{
    private readonly StateContainer _stateContainer;
    private readonly NavigationManager _navigationManager;
    private readonly ILogger<LoggingCircuitHandler> _logger;

    public LoggingCircuitHandler(StateContainer stateContainer, NavigationManager navigationManager, ILogger<LoggingCircuitHandler> logger)
    {
        _stateContainer = stateContainer;
        _navigationManager = navigationManager;
        _logger = logger;
    }

    public override Func<CircuitInboundActivityContext, Task> CreateInboundActivityHandler(Func<CircuitInboundActivityContext, Task> next)
    {
        return context =>
        {
            var scopeTags = new Dictionary<string, object>()
            {
                ["FormId"] = _stateContainer.FormId,
                ["BlazorUri"] = _navigationManager.Uri,
                ["CircuitId"] = context.Circuit.Id
            };

            using (_logger.BeginScope(scopeTags))
            {
                return next(context);
            }
        };
    }
}
```

**Comment by springy76** [comment]
Depending on the logger you could access scoped services by the use of async-locals initialized within `CreateInboundActivityHandler` created handler.

For example see https://learn.microsoft.com/en-us/aspnet/core/blazor/fundamentals/dependency-injection?view=aspnetcore-8.0#access-server-side-blazor-services-from-a-different-di-scope
(but IMHO the CircuitServicesAccessor there should be registered as singleton in order to do its thing, or the asynclocal be a static field somewhere)

Using Serilog for example you then would implement an `ILogEventEnricher` based enricher accessing this singleton at the time when log messages get created.

And you really should use `await next(context)` in your code (make the anonymous handler async).

**Reply by alexandrejobin** [reply]
> And you really should use `await next(context)` in your code (make the anonymous handler async).

I tried to make it async but the method itself is not async so I cannot do it. Or can you propose the new code I should use?

> Depending on the logger you could access scoped services by the use of async-locals initialized within CreateInboundActivityHandler created handler. Using Serilog for example you then would implement an ILogEventEnricher based enricher accessing this singleton at the time when log messages get created.

This is an idea that I would like to explore. But do you see something wrong about adding a scope to the my log the way I do it? That way, any log provider will work.

**Comment by jonathh21** [comment]
Shame this conversation went cold..

I too am trying to enrich my logs.

I haven't gotten as fat using dynamic data but 'BeginScope' in CreateInboundActivityHandler(Func<CircuitInboundActivityContext seems to work.

I am miffed as to why it doesn't also work if I use connection open, all of my understanding says it should if I scope the returned disposable as a field, but alas it doesn't.

If anyone knows why or if CreateInboundActivityHandler(Func<CircuitInboundActivityContext is a good / bad place to do this id appreciate it
